eos-grub-upgrade

```diff
@@ -0,0 +1,276 @@
+  if [[ -z ${esp_part} ]]; then
+    echo "ESP not found"
+    exit 1
+  fi
+
+  # On Endless-only (not dual boot) UEFI installation the ESP is always the first
+  # partition on the booting disk
+  local esp_partno=$(get_last_char "${esp_part}")
+  if [[ ${esp_partno} != "1" ]]; then
+    echo "ESP is not the first partition (how?)"
+    exit 1
+  fi
+
+  # Mount the ESP
+  if ! mount "${esp_part}" ${ESP_MOUNTPOINT}; then
+    echo "Failed to mount ESP"
+    exit 1
+  fi
+}
+
+check_for_upgrade() {
+  local old=$1
+  local new=$2
+
+  # Check both are regular files
+  if [[ -f ${old} && -f ${new} ]]; then
+
+    # pesign returns 0 when the file is signed, 1 otherwise (also when the file
+    # is not an UEFI application)
+    pesign -S -i "${new}" &> /dev/null
+    local new_signed=$?
+
+    pesign -S -i "${old}" &> /dev/null
+    local old_signed=$?
+
+    # If we are going to upgrade a signed file we must be sure that also the
+    # new file is signed
+    if [[ ${new_signed} -eq 0 && ${new_signed} -ne ${old_signed} ]]; then
+      return
+    fi
+
+    # Compare the two files, if they are different copy the new file into a
+    # temporary ${old}.new file in the same directory (and partition) of the
+    # file to overwrite
+    if ! cmp --silent "${old}" "${new}"; then
+      cp -a "${new}" "${old}".new
+      echo "${old}"
+    fi
+  fi
+}
+
+# Get the installation type. It could embed a complex logic to determine the
+# installation type.
+get_installation_type() {
+  local firmware_type=$1
+  local partition_table=$2
+
+  if [[ ${firmware_type} == "uefi" && ${partition_table} == "gpt" ]]; then
+    local inst_type
+    for inst_type in "${!GRUB_AMD64[@]}"; do
+      if [[ -f ${GRUB_AMD64[$inst_type]} ]]; then
+        echo "${inst_type}"
+        return
+      fi
+    done
+  fi
+}
+
+do_uefi_upgrade() {
+  local root_disk=$1
+  local old_file
+  local new_file
+
+  # Mount the ESP
+  mount_ESP "${root_disk}"
+
+  # Get the installation type
+  local inst_type=$(get_installation_type "uefi" "gpt")
+  if [[ -z ${inst_type} ]]; then
+    echo "Installation type unknown"
+    exit 1
+  fi
+  echo "Installation type: ($inst_type)"
+
+  GRUB_AMD64=${GRUB_AMD64[$inst_type]}
+  SHIM=${SHIM[$inst_type]}
+
+  # Check if the files need to be upgraded
+  local files_to_upgrade=( )
+  local f
+  for f in "${EFI_LIST[@]}"; do
+    eval old_file=\$"$f"
+    eval new_file=\$"NEW_$f"
+    local cur=$(check_for_upgrade "$old_file" "$new_file")
+    if [[ ! -z ${cur} ]]; then
+      echo "Preparing to upgrade: ${cur}"
+      files_to_upgrade+=( "${cur}" )
+    fi
+  done
+
+  if [[ "${#files_to_upgrade[@]}" -eq 0 ]]; then
+    echo "Nothing to upgrade"
+    return
+  fi
+
+  # Let's try to do this atomically using mv because the old and the new
+  # files are now in the same partition
+  local file
+  for file in "${files_to_upgrade[@]}"; do
+    mv -f "${file}.new" "${file}"
```

Comment:
What if this fails? Should we try to roll back any files that have already been overwritten?